src/installer/scripts.py

```diff
@@ -0,0 +1,116 @@
+        return None
+    name = "{}.exe".format(kind)
+    return importlib_resources.read_binary(_scripts, name)
+
+
+def _is_executable_simple(executable):
+    # type: (six.binary_type) -> bool
+    if b" " in executable:
+        return False
+    shebang_length = len(executable) + 3  # Prefix #! and newline after.
+    # According to distlib, Darwin can handle up to 512 characters. But I want
+    # to avoid platform sniffing to make this as platform agnostic as possible.
+    # The "complex" script isn't that bad anyway.
+    return shebang_length <= 127
+
+
+def _quote_compat(s):  # pragma: no cover
+    # type: (six.text_type) -> six.text_type
+    """Crude implementation taken from shlex.quote().
+    """
+    return u"'" + s.replace(u"'", u"'\"'\"'") + u"'"
+
+
+def _build_shebang(executable, kind):
+    # type: (six.text_type, LauncherKind) -> six.binary_type
+    """Build a shebang line.
+
+    The non-launcher cases are taken directly from distlib's implementation,
+    which tries its best to account for command length, spaces in path, etc.
+
+    https://bitbucket.org/pypa/distlib/src/58cd5c6/distlib/scripts.py#lines-124
+    """
+    executable_bytes = six.ensure_binary(executable, encoding="utf-8")
+    if kind:  # The launcher can just use the command as-is.
+        return b"#!" + executable_bytes
+    if _is_executable_simple(executable_bytes):
+        return b"#!" + executable_bytes
+
+    quote = getattr(shlex, "quote", _quote_compat)
+    quoted = six.ensure_binary(quote(executable), encoding="utf-8")
+    # I don't understand a lick what this is trying to do.
+    return b"#!/bin/sh\n'''exec' " + quoted + b' "$0" "$@"\n' + b"' '''"
+
+
+class Script(object):
+    __slots__ = ("name", "module", "attr")
+
+    def __init__(self, name, module, attr):
+        # type: (str, str, str) -> None
+        self.name = name
+        self.module = module
+        self.attr = attr
+
+    def __repr__(self):
+        # type: () -> str
+        return "Script(name={!r}, module={!r}, attr={!r}".format(
+            self.name, self.module, self.attr,
+        )
+
+    def build(self, executable, kind):
+        # type: (str, LauncherKind) -> Tuple[str, six.binary_type]
+        assert kind in _ALLOWED_LAUNCHER_KINDS, "{!r} not in {!r}".format(
+            kind, _ALLOWED_LAUNCHER_KINDS,
+        )
+
+        launcher = _get_launcher_data(kind)
+        shebang = _build_shebang(executable, kind)
+        code = _SCRIPT_TEMPLATE.format(
+            module=self.module,
+            import_name=self.attr.split(".")[0],
+            func_path=self.attr,
+        ).encode("utf-8")
+
+        if launcher is None:
+            return (self.name, shebang + b"\n" + code)
+
+        stream = io.BytesIO()
+        with zipfile.ZipFile(stream, "w") as zf:
+            zf.writestr("__main__.py", code)
+        name = "{}.exe".format(self.name)
+        data = launcher + shebang + b"\n" + stream.getvalue()
+        return (name, data)
```

Comment:
I wonder if there’s a way to stream the data instead to avoid building bytes in memory. In most scenarios people are going to write the return value directly to a file anyway.

Reply:
It still makes sense to do this IMO, but the API will be different (because `importlib.resources.open_binary()` is a context manager). So the two functions can co-exist, one (in this PR) returning `bytes` and the other a stream.

With that conclusion, I think all topics are covered for script generation, and this is ready to go in.